File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/ubicacion/UbicacionJerarquiaService.java
package net.latinus.sistema.integral.gestion.seguridad.service.ubicacion;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.BodyEncriptado;
import net.latinus.sistema.integral.gestion.seguridad.model.both.ubicacion.UbicacionJerarquiaDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;

import java.util.List;

/**
 * Interface de servicio para UbicacionJerarquia
 * Define los contratos para las operaciones de ubicaciones jerárquicas
 */
public interface UbicacionJerarquiaService {

    /**
     * Obtiene una lista paginada de ubicaciones jerárquicas
     *
     * @param httpServletRequest request de la petición
     * @param bodyEncriptado objeto encriptado con parámetros de paginación
     * @return RespuestaPorDefectoAuditoria con PaginacionResponse de UbicacionJerarquiaDTO
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<UbicacionJerarquiaDTO>> obtenerListaPaginada(
            HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);

    /**
     * Obtiene la lista completa de ubicaciones jerárquicas no removidas
     *
     * @param httpServletRequest request de la petición
     * @return RespuestaPorDefectoAuditoria con lista de UbicacionJerarquiaDTO
     */
    RespuestaPorDefectoAuditoria<List<UbicacionJerarquiaDTO>> obtenerListaCompleta(
            HttpServletRequest httpServletRequest);

    /**
     * Obtiene una ubicación jerárquica por su token identificador
     *
     * @param httpServletRequest request de la petición
     * @param tokenIdentificador token de identificación
     * @return RespuestaPorDefectoAuditoria con UbicacionJerarquiaDTO
     */
    RespuestaPorDefectoAuditoria<UbicacionJerarquiaDTO> obtenerPorTokenIdentificador(
            HttpServletRequest httpServletRequest, String tokenIdentificador);

    /**
     * Obtiene los hijos de una ubicación jerárquica padre
     *
     * @param httpServletRequest request de la petición
     * @param tokenIdentificadorPadre token del padre
     * @return RespuestaPorDefectoAuditoria con lista de UbicacionJerarquiaDTO
     */
    RespuestaPorDefectoAuditoria<List<UbicacionJerarquiaDTO>> obtenerHijosPorTokenIdentificadorPadre(
            HttpServletRequest httpServletRequest, String tokenIdentificadorPadre);

    /**
     * Obtiene ubicaciones jerárquicas por token identificador de jerarquía centro
     *
     * @param httpServletRequest request de la petición
     * @param tokenIdentificadorCentro token de la jerarquía centro
     * @return RespuestaPorDefectoAuditoria con lista de UbicacionJerarquiaDTO
     */
    RespuestaPorDefectoAuditoria<List<UbicacionJerarquiaDTO>> obtenerPorTokenIdentificadorJerarquiaCentro(
            HttpServletRequest httpServletRequest, String tokenIdentificadorCentro);

    /**
     * Crea o edita una ubicación jerárquica
     *
     * @param httpServletRequest request de la petición
     * @param bodyEncriptado objeto encriptado con datos de UbicacionJerarquiaDTO
     * @return RespuestaPorDefectoAuditoria con UbicacionJerarquiaDTO creada o editada
     */
    RespuestaPorDefectoAuditoria<UbicacionJerarquiaDTO> crearEditar(
            HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);

    /**
     * Elimina una ubicación jerárquica
     *
     * @param httpServletRequest request de la petición
     * @param bodyEncriptado objeto encriptado con token de identificación
     * @return RespuestaPorDefectoAuditoria con UbicacionJerarquiaDTO eliminada
     */
    RespuestaPorDefectoAuditoria<UbicacionJerarquiaDTO> eliminar(
            HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);
}

File: pronacej-front/src/app/modules/ubicacion/services/ubicacionJerarquia.service.ts
import { Injectable } from '@angular/core';
import { UbicacionJerarquiaDTO } from '../../../core/model/both/ubicacionJerarquiaDTO.model';
import { PaginacionRequest } from '../../../core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from '../../../core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from '../../../core/model/response/RespuestaPorDefecto.model';
import { BackendService } from '../../../core/services/backend.service';
import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class UbicacionJerarquiaService {
    private path = '/ubicacion-jerarquia';

    constructor(private backendService: BackendService) {}

    obtenerListaPaginada(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<UbicacionJerarquiaDTO>>> {
        const endPoint = this.path + '/obtenerListaPaginada';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerListaCompleta(
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<UbicacionJerarquiaDTO[]>> {
        const endPoint = this.path + '/obtenerListaCompleta';
        return this.backendService.getFinal(endPoint, {}, nemonicoMenu);
    }

    obtenerPorTokenIdentificador(
        tokenIdentificador: string,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<UbicacionJerarquiaDTO>> {
        const endPoint = this.path + '/obtenerPorTokenIdentificador';
        return this.backendService.getFinal(
            endPoint,
            { tokenIdentificador },
            nemonicoMenu
        );
    }

    obtenerHijosPorTokenIdentificadorPadre(
        tokenIdentificadorPadre: string,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<UbicacionJerarquiaDTO[]>> {
        const endPoint =
            this.path + '/obtenerHijosPorTokenIdentificadorPadre';
        return this.backendService.getFinal(
            endPoint,
            { tokenIdentificadorPadre },
            nemonicoMenu
        );
    }

    obtenerPorTokenIdentificadorJerarquiaCentro(
        tokenIdentificadorCentro: string,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<UbicacionJerarquiaDTO[]>> {
        const endPoint =
            this.path + '/obtenerPorTokenIdentificadorJerarquiaCentro';
        return this.backendService.getFinal(
            endPoint,
            { tokenIdentificadorCentro },
            nemonicoMenu
        );
    }

    crearEditar(
        ubicacionJerarquiaDTO: UbicacionJerarquiaDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<UbicacionJerarquiaDTO>> {
        const endPoint = this.path + '/crearEditar';
        return this.backendService.postFinal(
            endPoint,
            ubicacionJerarquiaDTO,
            nemonicoMenu
        );
    }

    eliminar(
        ubicacionJerarquiaDTO: UbicacionJerarquiaDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<UbicacionJerarquiaDTO>> {
        const endPoint = this.path + '/eliminar';
        return this.backendService.postFinal(
            endPoint,
            ubicacionJerarquiaDTO,
            nemonicoMenu
        );
    }

    async checkError(error: any, mostrarError = true): Promise<string> {
        return await this.backendService.checkError(error, mostrarError);
    }
}
